lib/web_ui/lib/src/engine/window.dart

```diff
@@ -168,6 +170,66 @@ class EngineWindow extends ui.Window {
     _onLocaleChangedZone = Zone.current;
   }
 
+  /// The locale used when we fails to get the list from the browser.
+  static const _defaultLocale = const ui.Locale('en', 'US');
+
+  /// We use the first locale in the [locales] list instead of the browser's
+  /// built-in `navigator.language` because browsers do not agree on the
+  /// implementation.
+  ///
+  /// See also:
+  ///
+  /// * https://developer.mozilla.org/en-US/docs/Web/API/NavigatorLanguage/languages,
+  ///   which explains browser quirks in the implementation notes.
+  @override
+  ui.Locale get locale => _locales.first;
+
+  @override
+  List<ui.Locale> get locales => _locales;
+  List<ui.Locale> _locales = parseBrowserLanguages();
+
+  /// Sets locales to `null`.
+  ///
+  /// `null` is not a valid value for locales. This is only used for testing
+  /// locale update logic.
+  void debugResetLocales() {
+    _locales = null;
+  }
+
+  // Called by DomRenderer when browser languages change.
+  void _updateLocales() {
+    _locales = parseBrowserLanguages();
+  }
+
+  static List<ui.Locale> parseBrowserLanguages() {
+    // TODO(yjbanov): find a solution for IE
+    if (!js_util.hasProperty(html.window.navigator, 'languages')) {
+      return const [_defaultLocale];
+    }
+
+    final List<ui.Locale> locales = <ui.Locale>[];
+    for (final String language in html.window.navigator.languages) {
+      final List<String> parts = language.split('-');
+      if (parts.length > 1) {
+        locales.add(ui.Locale(parts.first, parts.last));
+      } else {
+        locales.add(ui.Locale(language));
+      }
+    }
+
+    // To make it easier for the app code, let's not leave the locales list
+    // empty. This way there's fewer corner cases for apps to handle.
+    if (locales.isEmpty) {
+      locales.add(_defaultLocale);
+    }
```

Comment:
Suggestion: return early before the for loop.
```dart
if (!js_util.hasProperty(html.window.navigator, 'languages')) {
  return const [_defaultLocale];
}

if (html.window.navigator.languages.isEmpty) {
  return const [_defaultLocale];
}

for (final String language in html.window.navigator.languages) {
  // ...
}
return locales;
```

This is just a personal preference of mine. It makes reading the code easier when edge cases are handled at the beginning.

Reply:
Done. I actually prefer that too, but missed it this time.